File: week-03_05/tp/tp/tp7/read_file.f90
program read_file
  use MPI_F08
  implicit none
  integer, parameter                  :: nb_values=121
  integer(kind=MPI_OFFSET_KIND)       :: offset
  integer, dimension(nb_values)       :: values
  TYPE(MPI_Status)                    :: status
  integer                             :: rank,nb_bytes_integer
  TYPE(MPI_File)                      :: fh

  call MPI_INIT()
  call MPI_COMM_RANK(MPI_COMM_WORLD,rank)
  call MPI_FILE_SET_ERRHANDLER(MPI_FILE_NULL,MPI_ERRORS_ARE_FATAL)

  ! TODO: Open the file "data.dat" in read mode

  values(:)=0
  ! TODO: Read via explicit offsets, in individual mode

  open(unit=45,file="file_dei"//achar(48+rank)//".dat")
  write(unit=45,fmt='(I3)') values(:)
  close(unit=45)
  call MPI_BARRIER(MPI_COMM_WORLD)
  if (rank == 0) then
    print *, 'First read done'
  end if

  values(:)=0
  ! TODO: Read via shared file pointers, in collective mode

  open(unit=45,file="file_ppc"//achar(48+rank)//".dat")
  write(unit=45,fmt='(I3)') values(:)
  close(unit=45)
  call MPI_BARRIER(MPI_COMM_WORLD)
  if (rank == 0) then
    print *, 'Second read done'
  end if

  values(:)=0
  ! TODO: Read via individual file pointer, in individual mode

  open(unit=45,file="file_pii"//achar(48+rank)//".dat")
  write(unit=45,fmt='(I3)') values(:)
  close(unit=45)
  call MPI_BARRIER(MPI_COMM_WORLD)
  if (rang == 0) then
    print *, 'Third read done'
  end if

  values(:)=0
  ! TODO: Read via shared file pointers, in individual mode

  open(unit=45,file="file_ppi"//achar(48+rank)//".dat")
  write(unit=45,fmt='(I3)') values(:)
  close(unit=45)
  call MPI_BARRIER(MPI_COMM_WORLD)
  if (rang == 0) then
    print *, 'Fourth read done'
  end if

  ! TODO: Close the file

  call MPI_FINALIZE()
end program read_file

File: week-03_05/tp/tp_C/tp7/read_file.c
#include <mpi.h>
#include <stdlib.h>
#include <stdio.h>

int main(int argc, char * argv[]) {
  const int nb_values=121;
  int values[nb_values];
  int rank,iter;
  MPI_File fh;
  int nb_bytes_integer;
  MPI_Offset offset;
  MPI_Status status;
  char name_file[256];
  FILE * file;

  MPI_Init( &argc, &argv);
  MPI_Comm_rank(MPI_COMM_WORLD,&rank);
  MPI_File_set_errhandler(MPI_FILE_NULL,MPI_ERRORS_ARE_FATAL);

  /* TODO: Open data.dat in read mode */

  for (iter=0;iter<nb_values; iter++) values[iter]=0;
  /* TODO: Read via explicit offsets, in individual mode */

  sprintf(name_file,"file_dei%1d.dat",rank);
  file = fopen(name_file,"w");
  for (iter=0; iter<nb_values; iter++)
    fprintf(file,"%3d\n",values[iter]);
  fclose(file);
  MPI_Barrier(MPI_COMM_WORLD);
  if (rank == 0)
    printf("First read done\n");

  for (iter=0;iter<nb_values; iter++) values[iter]=0;
  /* TODO: Read via shared ifle pointers, in collective mode */

  sprintf(name_file,"file_ppc%1d.dat",rank);
  file = fopen(name_file,"w");
  for (iter=0; iter<nb_values; iter++)
    fprintf(file,"%3d\n",values[iter]);
  fclose(file);
  MPI_Barrier(MPI_COMM_WORLD);
  if (rank == 0)
    printf("Second read done\n");

  for (iter=0;iter<nb_values; iter++) values[iter]=0;
  /* TODO: Read via individual file pointer, in individual mode */

  sprintf(name_file,"file_pii%1d.dat",rank);
  file = fopen(name_file,"w");
  for (iter=0; iter<nb_values; iter++)
    fprintf(file,"%3d\n",values[iter]);
  fclose(file);
  MPI_Barrier(MPI_COMM_WORLD);
  if (rank == 0)
    printf("Third read done\n");

  for (iter=0;iter<nb_values; iter++) values[iter]=0;
  /* TODO: Read via shared file pointers, in individual mode */

  sprintf(name_file,"file_ppi%1d.dat",rank);
  file = fopen(name_file,"w");
  for (iter=0; iter<nb_values; iter++)
    fprintf(file,"%3d\n",values[iter]);
  fclose(file);
  MPI_Barrier(MPI_COMM_WORLD);
  if (rank == 0)
    printf("Fourth read done\n");

  /* TODO: Close the file */

  MPI_Finalize();
  return 0;
}
